### PR TITLE
pbt: convert gpb-eqc to PropEr & add to test suite

### PR DESCRIPTION
**Pull request by fenollp**
This is a starting point for more thorough testing.
I intend on extending https://github.com/tomas-abrahamsson/gpb-eqc
with PropEr's Targeted testing properties.

### CODE REVIEW

**Comment by tomas-abrahamsson** [comment]
Noce! I'll be pushing a few additions. Rudimentary Makefile support because I find I like `make`, fitting the code to 80 columns (re-introducing the `?f(Fmt,Args)` but as a macro, to still get compiler warnings if the formatting string or args are wrong), and maybe a few more cleanups. As far as I could see, it looks like all functionality from `gpb-eqc` is still there, although a bit reworded at some places.

Among the changes are also a check for OTP >= 21.0 instead of exactly 21.0, hope travis likes my shell script, but knowing how difficult it can be to get all details right, I guess might have to push more than once.

**Comment by fenollp** [comment]
LGTM. Do you need anything from me?

**Comment by tomas-abrahamsson** [comment]
I plan to include this and some more in an upcoming release. Would you know by the way, is there a way with rebar to make it run all tests except prop_gpb? About the travis job: why not run these tests for more versions of Erlang?

**Comment by fenollp** [comment]
> is there a way with rebar to make it run all tests except prop_gpb

There is unfortunately no option!

> why not run these tests for more versions of Erlang

No reason to. Running against more versions of protobuf is needed though (`Vsn >= [3,0]`)

**Comment by fenollp** [comment]
For protobuf v3.* this can work:
```shell
protoc() {
    docker run -it --rm -v "$PWD":/w -w /w alpine /bin/sh -c 'apk update && apk upgrade && apk add protobuf && protoc '"$@"
}
```

**Comment by tomas-abrahamsson** [comment]
(I added #144 to track update libprotobuf.)

I'm a bit concerned with two issues:
* Missing option to run rebar{2,3} without proper tests. Since it takes some time to run, I'd actually prefer to have it as an option to run the proper tests, than an option to run without. I was thinking since rebar.config.script is a script, there could be a possibility to check some environment variable. Would this work, and what do you think?
* Unit tests don't work with rebar2 any more (I know of at least one that's still on rebar2 and that can't upgrade quickly). (and yes, I know there are some failing tests with rebar2 due to directory structures, but now I get `test/prop_gpb.erl:9: can't find include lib "proper/include/proper.hrl"` and no tests run at all) I tried `rebar get-deps` but nothing happens. Can this be worked around in the rebar.config.script? (I liked the case with rebar3 that it fetches the deps only when running `eunit`, not with `compile` which is what most gpb users do.)

File: .gitignore [hunk 1]
```diff
@@ -23,3 +23,7 @@ _build
 /descr_src/gpb_descriptor.hrl
 # Plt generated by Makefile
 /.gpb-*.plt
+# proper-test target download dir in the Makefile
+/.deps
+# when proper is fetched as a rebar dependency
+/rebar.lock
```
[the remark this answers is not in the document]
Comment (tomas-abrahamsson) [reply .gitignore]: Should it be put under version control instead? (honest question)

**Comment by tomas-abrahamsson** [comment]
> How about `rebar3 eunit $(ls test/gpb*_tests.erl | sed 's%test/%-m %g;s%.erl%%')`

Hehe, seriously?

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (fenollp) [reply .gitignore]: I haven’t encountered a case where it shouldn’t be. It helps catch mistakes in any case.

**Comment by tomas-abrahamsson** [comment]
> I love how much fixes this PR brings in!

:) Argh, it seems I was in the wrong git worktree when I pushed an update to the branch yesterday, and I accidentally included a merge to master that also contained some other work. Sorry.

**Comment by tomas-abrahamsson** [comment]
Hi, I thought I should give some feedback on my intentions for this PR.  (I've been silent for far too long, sorry about that; difficulties finding enough time.)

I'd like to first merge and push a few branches I've had boiling locally for some time (one of which was accidentally pushed here). Then I'd like to tend to this PR. I'm for including it, but I want
* rebar2 support still for some time
* the quickcheck/proper tests to somehow not run by default for a few reasons: taking long time, adding a (test-) dependency (gpb is otherwise free of dependencies).

About the second point: got to check if one could use rebar3 profiles to achieve it somehow, eg `rebar3 as qc eunit` or something.

First steps would anyway be to squash or amend all the various fixups and undo the accidental merge. Related is also that I have (only locally still) an improvement to gpb-eqc, to quickcheck flat oneofs. Will have to include that one here too.

One reason I haven't had too much of a hurry on this is also that I locally have a Jenkins flow that runs the gpb-eqc checks before I push. Still, I agree of course that it'd be a good thing to have included out of the box. (I do not (yet) locally have checks with various versions of protobuf, only the 3.0.0, so #144 is a little important.)

I'll write more when I start working on it again.

**Comment by fenollp** [comment]
These are valid points, and true I have been anxious for this PR to get merged :) but no rush really!

* What doesn't bode well with `rebar2` in this PR? I have no idea...
* `prop_gpb`'s entry point right now is `props_test_/0` which makes it part of eunit tests. Instead this can be removed and [rebar3_proper](https://hex.pm/packages/rebar3_proper) can be used, probably with [the new options syntax](https://github.com/ferd/rebar3_proper#per-properties-meta-functions). You can uncomment that dep in rebar.config. With this change only eunit tests would be run with today's code and `rebar3 as test proper` will only run the PropEr tests.

**Comment by tomas-abrahamsson** [comment]
Iirc, I think what goes wrong with rebar2 is that proper isn't recognized as a dep and then the eunit tests will fail. (But memory fades quickly so I'm not 100% sure this was it.) I'd find it ok if PropEr-tests could be easily run with rebar3. Making it easy and optional to run them with rebar2 would not be high prio. But they mustn't fail with rebar2.

Thanks for the pointer to rebar_proper, I'll be looking into it.

**Comment by tomas-abrahamsson** [comment]
I just created #151 as a continuation of this PR. I rebased it, but I don't think it is possible for me to rebase your PR, so I created a new. If it is ok with you, we close this PR and continue the discussion in the #151. Or just merge it if you're ok with how it currently looks (I'll squash the fixups first, of course)

**Comment by tomas-abrahamsson** [comment]
Included in 4.3.2, hence closing.

**Comment by tomas-abrahamsson** [comment]
FYI: I'll unfortunately revert this (or the #151 merge, actually) due to the licensing issues brought up in #152.

**Pull request closed without merging**